### PR TITLE
chore: add automated CI testing with —no-intl node

### PR DESCRIPTION
**Pull request by mweberxyz**
## This relates to...

- https://github.com/nodejs/undici/issues/3002
- https://github.com/nodejs/undici/pull/2999
- https://github.com/nodejs/node/pull/52175

## Rationale

This PR extends the Node CI workflow to run undici tests against Node 20 and 21 built with `--without-intl` flag, which would have prevented the regression fixed by https://github.com/nodejs/undici/pull/2999 before the commit even merged to undici/main.

The full test suite currently failing against `main`, so problematic tests have been fixed where possible, otherwise excluded from running in the w/o intl build. If possible, these problematic tests/functionality should be fixed to run in w/o intl builds and the tests added to `test:javascript:withoutintl` (or even remove it entirely and both the w/o intl and regular builds can use `test:javascript`).

Note that the initial run, and occasional runs thereafter, of the workflow will take a long time (around 1 hour for the clean Node 21 build) -- but subsequent runs will be much faster due to ccache caching.

## Changes
- Workflows:
  - add `test-without-intl` job to workflows/nodejs.yml
- package.json:
  - add `test:javascript:withoutintl` script to run tests known good with an w/o intl node build
  - add `test:eventsource:nobuild`, `test:fetch:nobuild` scripts to pull test running logic out of `test:eventsource` and `test:fetch` because undici is built _prior_ to running w/o intl tests
  - add `test:wpt:withoutintl` script to run wpt tests except those problematic for w/o intl, namely:
    - `test/wpt/start-websockets.mjs`, and
    - `test/wpt/start-FileAPI.mjs`
 - various test tweaks to keep existing tests running without disabling for w/o intl

### Features

N/A

### Bug Fixes

N/A

### Breaking Changes and Deprecations

N/A

## Status

- [x] I have read and agreed to the [Developer's Certificate of Origin][cert]
- [x] Tested
- [ ] Documented
- [x] Review ready
- [ ] In review
- [ ] Merge ready

[cert]: https://github.com/nodejs/undici/blob/main/CONTRIBUTING.md

### CODE REVIEW

**Comment by mweberxyz** [comment]
I started looking into the actual test failures, the problems might not be all that arduous to fix and get the tests passing.

First, there's a failing test which is simply a misaligned console.table assertion.

Then, some tests run `npm run build:node` which fails, but maybe that doesn't maybe doesn't need to run in this configuration because the package can be built with a regular node?

I've gotten down into the `test:wpt` which is failing on steps that are clearly unicode related:
CleanShot 2024-03-28 at 21 56 56@2x

Will update with more findings and if I get the tests to fully pass.

**Comment by KhafraDev** [comment]
that failure will be ignored once #2728 lands cc @ronag @mcollina

**Comment by mweberxyz** [comment]
With most recent commits, w/o intl build tests are green, PR description has been updated, and ready for review.

As a example, I reintroduced the commit which caused nodejs CI to fail its w/o intl build, and the check failed as expected -- see https://github.com/mweberxyz/nodejs-undici/pull/6

File: test/node-test/debug.js [hunk 1]
```diff
@@ -9,6 +9,9 @@ const { tspl } = require('@matteo.collina/tspl')
 const removeEscapeColorsRE = /[\u001b\u009b][[()#;?]*(?:[0-9]{1,4}(?:;[0-9]{0,4})*)?[0-9A-ORZcf-nqry=><]/g
 
 test('debug#websocket', async t => {
+  if (!process.versions.icu) {
+    return
+  }
```
Comment (metcoder95) [inline test/node-test/debug.js:14]:
```suggestion
test('debug#websocket', { skip: !process.versions.icu }, async t => {
```

File: lib/mock/pending-interceptors-formatter.js [hunk 2]
```diff
@@ -7,6 +7,9 @@ const { Console } = require('node:console')
  * Gets the output of `console.table(…)` as a string.
  */
 module.exports = class PendingInterceptorsFormatter {
+  static PERSISTENT = process.versions.icu ? '✅' : 'Y '
+  static NOT_PERSISTENT = process.versions.icu ? '❌' : 'N '
+
```
Comment (Uzlopak) [inline lib/mock/pending-interceptors-formatter.js:12]: Do we need to add these static properties?
Cant we define them outside the class?

File: test/mock-interceptor-unused-assertions.js [hunk 3]
```diff
@@ -10,6 +10,10 @@ const util = require('../lib/core/util')
 // https://github.com/nodejs/node/pull/50135
 const tableRowsAlignedToLeft = util.nodeMajor >= 21 || (util.nodeMajor === 20 && util.nodeMinor >= 11)
 
+// `console.table` treats emoji as two character widths for cell width determination
+const Y = PendingInterceptorsFormatter.PERSISTENT
+const N = PendingInterceptorsFormatter.NOT_PERSISTENT
```
Comment (Uzlopak) [inline test/mock-interceptor-unused-assertions.js:15]:
```suggestion
const Y = process.versions.icu ? '✅' : 'Y '
const N = process.versions.icu ? '❌' : 'N '
```

File: package.json [hunk 4]
```diff
@@ -69,10 +69,13 @@
     "lint:fix": "standard --fix | snazzy",
     "test": "npm run test:javascript && cross-env NODE_V8_COVERAGE=  npm run test:typescript",
     "test:javascript": "node scripts/generate-pem && npm run test:unit && npm run test:node-fetch && npm run test:fetch && npm run test:cookies && npm run test:eventsource && npm run test:wpt && npm run test:websocket && npm run test:node-test && npm run test:jest",
+    "test:javascript:withoutintl": "node scripts/generate-pem && npm run test:unit && npm run test:node-fetch && npm run test:fetch:nobuild && npm run test:cookies && npm run test:eventsource:nobuild && npm run test:wpt:withoutintl && npm run test:node-test",
     "test:cookies": "borp -p \"test/cookie/*.js\"",
     "test:node-fetch": "borp -p \"test/node-fetch/**/*.js\"",
-    "test:eventsource": "npm run build:node && borp --expose-gc -p \"test/eventsource/*.js\"",
-    "test:fetch": "npm run build:node && borp --expose-gc -p \"test/fetch/*.js\" && borp -p \"test/webidl/*.js\" && borp -p \"test/busboy/*.js\"",
+    "test:eventsource": "npm run build:node && npm run test:eventsource:nobuild",
```
Comment (Uzlopak) [inline package.json:75]: why is building bad?

Comment (mweberxyz) [reply package.json:75]: Build fails with icu-less node build:

CleanShot 2024-03-29 at 14 25 05@2x

But we shouldn't to support _building_ undici with icu-less node, only running it.

This is why the workflow [installs regular node then builds undici](https://github.com/nodejs/undici/pull/3015/files#diff-e98936aa52a6dd7416e4296e9628456227d834f7245967383fd9ff80fd985dadR88) prior to switching to the icu-less build for running tests.

I tried to minimize changes to package.json, so as to make people not need to think about keeping the `test:javascript:withoutintl` script up-to-date with test strategy changes.

File: lib/mock/pending-interceptors-formatter.js (hunk 2 above)
Comment (mweberxyz) [reply lib/mock/pending-interceptors-formatter.js:12]: I imported them from test file, and given that the default export is a class, static members was cleanest. Seeing your next comment, I'll pull them up out of the class here, and re-define in test.

**Comment by mweberxyz** [comment]
One other omission I realize I haven't mentioned: `jest` doesn't work with icu-less node:

CleanShot 2024-03-29 at 14 55 14@2x

**Comment by mweberxyz** [comment]
@fraxken Does anything in here conflict with your auditing re: #3012? I tried to be more mindful of security implications this time around (ie- I reviewed the code for [this newly called action](https://github.com/nodejs/undici/pull/3015/files#diff-e98936aa52a6dd7416e4296e9628456227d834f7245967383fd9ff80fd985dadR118) and made sure to pin it), but would appreciate an extra set of eyes.

**Comment by fraxken** [comment]
@mweberxyz seems ok to me 😉

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by mcollina** [comment]
Could this be landed?

**Review by metcoder95 — APPROVED** [review]

**Pull request merged**